src/Compilers/Core/Portable/ExtendedSpecialType.cs

```diff
@@ -0,0 +1,70 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+
+namespace Microsoft.CodeAnalysis
+{
+    /// <summary>
+    /// A structure meant to represent a union of <see cref="SpecialType"/> and <see cref="InternalSpecialType"/> values
+    /// </summary>
+    [StructLayout(LayoutKind.Explicit, Size = 1)]
+    internal readonly struct ExtendedSpecialType
+    {
+        [FieldOffset(0)]
+        private readonly sbyte _value;
+
+        private ExtendedSpecialType(int value)
+        {
+            Debug.Assert(value >= sbyte.MinValue && value <= sbyte.MaxValue);
+            _value = (sbyte)value;
+        }
+
+        public static implicit operator ExtendedSpecialType(SpecialType value) => new ExtendedSpecialType((int)value);
+        public static explicit operator SpecialType(ExtendedSpecialType value) => value._value < (int)InternalSpecialType.First ? (SpecialType)value._value : SpecialType.None;
+
+        public static implicit operator ExtendedSpecialType(InternalSpecialType value) => new ExtendedSpecialType((int)value);
+
+        public static explicit operator ExtendedSpecialType(int value) => new ExtendedSpecialType(value);
+        public static explicit operator int(ExtendedSpecialType value) => value._value;
+
+        public static bool operator ==(ExtendedSpecialType left, SpecialType right) => left._value == (byte)right;
+        public static bool operator !=(ExtendedSpecialType left, SpecialType right) => !(left == right);
+
+        public static bool operator ==(SpecialType left, ExtendedSpecialType right) => right == left;
+        public static bool operator !=(SpecialType left, ExtendedSpecialType right) => right != left;
+
+        public static bool operator ==(ExtendedSpecialType left, ExtendedSpecialType right) => left._value == right._value;
+        public static bool operator !=(ExtendedSpecialType left, ExtendedSpecialType right) => !(left == right);
+
+        public override bool Equals(object? obj)
+        {
+            switch (obj)
+            {
+                case ExtendedSpecialType other:
+                    return this == other;
+
+                case SpecialType other:
+                    return this == other;
+
+                case InternalSpecialType other:
+                    return this == other;
+
+                default:
+                    return false;
+            }
+        }
+
+        public override int GetHashCode()
+        {
+            return _value.GetHashCode();
+        }
+
+        public override string ToString()
+        {
+            return _value.ToString();
```

Comment:
I'm wondering whether it would be useful, for debugging purposes, to instead have `ToString` determine whether `_value` is a special type or an internal special type, then call `ToString` on the appropriate value, so we get the right string representation. I think I'd at least like a `DebuggerDisplay` with that logic.